oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientProviderBuilder.java

```diff
@@ -0,0 +1,267 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client;
+
+import org.springframework.security.oauth2.client.endpoint.OAuth2AccessTokenResponseClient;
+import org.springframework.security.oauth2.client.endpoint.OAuth2ClientCredentialsGrantRequest;
+import org.springframework.security.oauth2.client.endpoint.OAuth2RefreshTokenGrantRequest;
+import org.springframework.util.Assert;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * A builder that builds a {@link DelegatingOAuth2AuthorizedClientProvider} composed of
+ * one or more {@link OAuth2AuthorizedClientProvider}(s) that implement specific authorization grants.
+ * The supported authorization grants are {@link #authorizationCode() authorization_code},
+ * {@link #refreshToken() refresh_token} and {@link #clientCredentials() client_credentials}.
+ * In addition to the standard authorization grants, an implementation of an extension grant
+ * may be supplied via {@link #provider(OAuth2AuthorizedClientProvider)}.
+ *
+ * @author Joe Grandja
+ * @since 5.2
+ * @see OAuth2AuthorizedClientProvider
+ * @see AuthorizationCodeOAuth2AuthorizedClientProvider
+ * @see RefreshTokenOAuth2AuthorizedClientProvider
+ * @see ClientCredentialsOAuth2AuthorizedClientProvider
+ * @see DelegatingOAuth2AuthorizedClientProvider
+ */
+public final class OAuth2AuthorizedClientProviderBuilder {
+	private final Map<Class<?>, Builder> builders = new HashMap<>();
+
+	private OAuth2AuthorizedClientProviderBuilder() {
+	}
+
+	/**
+	 * Returns a new {@link OAuth2AuthorizedClientProviderBuilder} for configuring the supported authorization grant(s).
+	 *
+	 * @return the {@link OAuth2AuthorizedClientProviderBuilder}
+	 */
+	public static OAuth2AuthorizedClientProviderBuilder withProvider() {
+		return new OAuth2AuthorizedClientProviderBuilder();
+	}
+
+	/**
+	 * Configures an {@link OAuth2AuthorizedClientProvider} to be composed with the {@link DelegatingOAuth2AuthorizedClientProvider}.
+	 * This may be used for implementations of extension authorization grants.
+	 *
+	 * @return the {@link OAuth2AuthorizedClientProviderBuilder}
+	 */
+	public OAuth2AuthorizedClientProviderBuilder provider(OAuth2AuthorizedClientProvider provider) {
+		Assert.notNull(provider, "provider cannot be null");
+		this.builders.computeIfAbsent(provider.getClass(), k -> () -> provider);
```

Comment:
Do you think a user would do this?

```
OAuth2AuthorizedClientProvider authorizedClientProvider =
		OAuth2AuthorizedClientProviderBuilder.withProvider()
				.authorizationCode()
				.refreshToken()
				.clientCredentials()
				.clientCredentials()     // The previous (first) instance is returned
				.build();
```

If if they did, they are able to override the configuration of the first `ClientCredentialsGrantBuilder` so I don't see an issue here.

Reply:
Agreed that there is little reason to call `clientCredentials` or other related methods twice.

What this reminds me of, though, is `JwtConfigurer`, which has both `jwtDecoder()` and `jwkSetUri()`. Whichever is called last is the one that takes precedence. In my mind:

`jwtDecoder` : `provider` and
`jwkSetUri` : `authorizationCode`

Hypothetically, if an API were to return an instance of `OAuth2AuthorizedClientProviderBuilder` (like `ClientRegistrations` returns a `ClientRegistraions.Builder`), I don't see how an application could override the configuration performed by that API.

For example, an application can do `ClientRegistrations.fromIssuerLocation(location).jwkSetUri(...)` to override what is provided by `ClientRegistrations` since `ClientRegistration.Builder` takes the last configuration as the correct one.

Reply:
I see the examples you provided but just to re-clarify, if a user calls `clientCredentials()` a second (or more) time it will return the same instance each time:

1st time
`clientCredentials(configurer -> configurer.accessTokenResponseClient(accessTokenResponseClient))`

2nd time
`clientCredentials(configurer -> configurer.accessTokenResponseClient(accessTokenResponseClient))`

The 2nd call overrides the previous `accessTokenResponseClient` set so overriding works  here.

Am I missing something or not understanding you last comment?

Reply:
Okay, I see. Since the configurer method returns the same instance, overriding works. This was my main concern with doing first-wins.